rust-runtime/smithy-http/src/event_stream.rs

```diff
@@ -0,0 +1,314 @@
+
+use crate::body::SdkBody;
+use crate::result::SdkError;
+use bytes::Bytes;
+use bytes_utils::SegmentedBuf;
+use hyper::body::HttpBody;
+use smithy_eventstream::frame::{
+    DecodedFrame, MarshallMessage, MessageFrameDecoder, SignMessage, UnmarshallMessage,
+};
+use std::error::Error as StdError;
+use std::marker::PhantomData;
+use std::ops::DerefMut;
+
+enum SenderState {
+    /// A Hyper body channel has been created for the Event Stream input,
+    /// but it's not hooked up to anything yet.
+    NotReady(hyper::Body),
+    /// The Hyper body channel is all wired up, and we have a message signer implementation.
+    Ready(Box<dyn SignMessage>),
+}
+
+/// Sends Smithy-modeled messages on an Event Stream.
+pub struct Sender<T, E: StdError + Send + Sync> {
+    marshaller: Box<dyn MarshallMessage<Input = T>>,
+    http_sender: hyper::body::Sender,
+    state: SenderState,
+    _phantom: PhantomData<E>,
+}
+
+impl<T, E: StdError + Send + Sync> Sender<T, E> {
+    /// Creates a new `Sender` with the given message marshaller.
+    pub fn new(marshaller: impl MarshallMessage<Input = T> + 'static) -> Self {
+        let (http_sender, http_body) = hyper::Body::channel();
+        Sender {
+            marshaller: Box::new(marshaller),
+            http_sender,
+            state: SenderState::NotReady(http_body),
+            _phantom: Default::default(),
+        }
+    }
+
+    fn signer_mut(&mut self) -> &mut dyn SignMessage {
+        match &mut self.state {
+            SenderState::NotReady(_) => panic!("Sender is not ready to send data yet"),
+            SenderState::Ready(ref mut state) => state.deref_mut(),
+        }
+    }
+
+    /// Asynchronously sends a message on the stream. Returns `Ok(())` if the message was sent
+    /// successfully. Returns `Err(SdkError::ConstructionFailure)` if the message couldn't be
+    /// marshalled into an Event Stream frame, (e.g., if the message payload was too large).
+    /// Returns `Err(SdkError::DispatchFailure)` if a transport failure occurred.
+    pub async fn send(&mut self, input: T) -> Result<(), SdkError<E>> {
+        let message = self
+            .marshaller
+            .marshall(input)
+            .map_err(|err| SdkError::ConstructionFailure(Box::new(err)))?;
+        let message = self
+            .signer_mut()
+            .sign(message)
+            .map_err(|err| SdkError::ConstructionFailure(err))?;
+        let mut buffer = Vec::new();
+        message
+            .write_to(&mut buffer)
+            .map_err(|err| SdkError::ConstructionFailure(Box::new(err)))?;
+        self.http_sender
+            .send_data(buffer.into())
+            .await
+            .map_err(|err| SdkError::DispatchFailure(Box::new(err)))
+    }
+
+    // Intended to be used by the generated code to connect the `Sender` to the actual request.
+    #[doc(hidden)]
+    pub fn make_ready(&mut self, signer: impl SignMessage + 'static) -> hyper::Body {
```

Comment:
this feels like a builder? I'd probably try to avoid this enum state pattern if possible. What forces the signer to be bound in later?

Reply:
Woops, I forgot that we wanted to take a stream trait in the input, so I was heading down the path of making Sender work as follows:

```rust
let my_sender = Sender::new();

let output = client.some_event_stream_operation()
    .something(my_sender)
    .call()
    .await?;

my_sender.send(SomeMessage::builder().foo(5).build());
```

This had its own issues. For example, the builder will need to take a mutable reference to the sender, or alternatively, the sender would need to have an `Arc` inner so that it can be cloned and mutated.

Using the stream trait will side-step the entire problem though, so I'll course correct :blush: